docs/kb/rules/rule-file-perf-principles.md

```diff
@@ -0,0 +1,62 @@
+---
+description: Learn the rule and file performance principles to abide by when scanning repositories to optimize scan times.
+tags:
+  - Rules
+  - Semgrep Registry
+  - Semgrep Code
+---
+
+import MoreHelp from "/src/components/MoreHelp"
+
+# Performance principles for rules and files to abide by when scanning repositories
+
+## Rules
+
+The amount of time required for rules to run scales better than linearly when
+adding interfile rules, which are those with `interfile: true` in the options.
+That is, doubling the number of interfile rules increases the runtime, but not
+by double. However, some rules run faster than others, and adding a slow rule
+when all the rest are fast can cause a significant slowdown.
+
+Rules are slower if the sub-patterns, such as `pattern: <... $X ...>`, result in
+a greater number of matches. When writing rules, pay special attention to the
+problems raised by sub-pattern matches. The most important factor for runtime is
+the time spent adding to various lists/sets.
+
+You can benchmark your rules by adding the `--time` flag to your `semgrep scan`
+command. When you use this flag, your results return with a timing summary; if
+your output format is JSON, you'll see times for each rule-target pair.
+
+## Files
+
+Generally, the time required to scan files scales linearly with the number of
+files scanned, but file size is still important. Overall, the time taken is
+**time for setup work + time for matching**. For setup work, files aren’t
+analyzed alone but in groups of mutually dependent files (SCCs).
```

Comment:
@emjin what does SCC stand for?

Reply:
SCC stands for  strongly connected components, yes @emjin

Reply:
That's correct!